crates/ironrdp-async/src/connector.rs

```diff
@@ -68,17 +87,99 @@ where
     Ok(result)
 }
 
+async fn resolve_generator(
+    generator: &mut CredSspProcessGenerator<'_>,
+    mut network_client: Box<dyn AsyncNetworkClient>,
```

Comment:
This is my initial thoughts as well, but since we passed the network down the stack as Option<&mut AsyncNetworkClient> and we use it in a loop, we cannot move out of options.

Reply:
if this is not desired way of doing it, we can try consume the network client at top level. Let me know which way you perfer

Reply:
Oh yeah, the error is not straightforward. Actually, stuff like `Option<&mut T>` are not super ergonomic in some cases. However, you can definitely retrieve the `&mut T` inside.
```rust
async fn perform_credssp_step<S>(
    /* … */
    mut network_client: Option<&mut dyn AsyncNetworkClient>, // <- make the binding mutable
    /* … */
) -> ConnectorResult<()>
where
    S: FramedRead + FramedWrite,
{
    /* … */
            if let Some(network_client) = network_client.as_deref_mut() { // <- use `.as_deref_mut()`
                trace!("resolving network");
                resolve_generator(&mut generator, network_client).await?
            } else {
                generator
                    .resolve_to_result()
                    .map_err(|e| custom_err!(" cannot resolve generator without a network client", e))?
            }
    /* … */
```

This way, you can borrow "exclusively" the `Option<&mut T>`, ensuring no one else have access to the inner `&mut T` at the same time, without moving `&mut T` out of it (you should think of `&mut` as "exclusive reference" in Rust).

Also you can basically forget about the `ref` keyword. It’s not useful since https://rust-lang.github.io/rfcs/2005-match-ergonomics.html (you may see it in old codebases, but that’s pretty much about it)

Reply:
updated